### ISSUE TITLE
Some problems about parameter setting

### ISSUE THREAD

**Issue by baobaba13** [ComputationalRadiationPhysics/picongpu#4190]
Hi dear friends
Use the OMP backend to simulate whether there is parallel computing. If it can, how to set parallelism? I haven't found a way to set it. In addition, in order to better understand the example "bremsstrahlung: emission of bremsstrahlung from laser foot interaction", I read the paper "Entwicklung und Ü berpr ü Fung eines photonmodels f ü r die abstractlung durch hochenergetische elektronen". I have the following questions to ask:

1. How to set the plasma density `ne` in the `.Param` file.
![1](https://user-images.githubusercontent.com/39699124/176632623-24a910c5-3667-40bb-97b8-0768f63ee563.jpg)

2. How to set the atomic number `Z` in the `.Param` file.
![2](https://user-images.githubusercontent.com/39699124/176632661-18c8fc66-a943-46b2-8864-7e0463b7c2fe.jpg)

3. Is code a in the `particle.param` file（`static constexpr uint32_t numParticlesPerCell = 100u;`） used to set the number of macro particles in each cell? If so, can this `100` be set smaller according to the literature?
![3](https://user-images.githubusercontent.com/39699124/176632705-12bdd37f-1215-444e-8b65-ffcb1dc3f47d.jpg)

**Comment by sbastrakov** [comment]
Hello @baobaba13 ,

Regarding OpenMP, please follow [this documentation page](https://picongpu.readthedocs.io/en/latest/install/profile.html#general-notes-on-using-cpus). Potentially, the [next section](https://picongpu.readthedocs.io/en/latest/install/profile.html#your-workstation) can also be useful for you.

For your other questions,

1. In `.param` files density is measured in units of `BASE_DENSITY` set in `density.param`. All preset density profiles operate with that value. So e.g. `Homogenous` profile means uniform distribution with that density, etc. One normally generates macroparticles according to a chosen density profile at initialization, as expressed in `speciesInitialization.param` with `CreateDensity<...>`. A species can define its own `densityRatio` to express that it should use `BASE_DENSITY` times that value as its density. Perhaps this is what you need, please see `FoilLCT` example.

2. I think you are looking for [this](https://github.com/ComputationalRadiationPhysics/picongpu/blob/829c105b5be8c7ab490d319aaaed4ebabfb2d43f/include/picongpu/param/ionizer.param#L178), please see how it is used in our examples, e.g. `FoilLCT`.

3. Yes, this is number of macroparticles per cell, you could change it. Note that this number is independent from density - for any given density and number of macroparticles, `CreateDensity<>` will match the macroparticle weights to represent the given density with given number of macroparticles.

**Comment by baobaba13** [comment]
Hi @sbastrakov According to your reply, I know that picongpu can use MPI parallelism.  The build option what I use is always `-b omp2b:native.` Can I understand the meaning of this option as running simulation on only one computing node? My CPU model is shown in the following figure. According to PIC build `-b omp2b:znver3,` if I want to use parallelism, can I simply modify this option to achieve the purpose of parallelism? If not, could you give me some detailed guidance? Thank you very much!
![QQ图片20220630203733](https://user-images.githubusercontent.com/39699124/176679194-dae186e3-db67-4490-8119-f2708b792d30.png)

**Comment by sbastrakov** [comment]
These are different and almost unrelated things. MPI parallelization is independent from how each single CPU or GPU is used. How many MPI processes you will run is controlled by `-d` command-line option, or in your `.cfg` file by variables `TBG_devices_{x,y,z}`. Independently of that, inside each device, be it a CPU or a GPU, PIConGPU may also run in parallel using OpenMP, CUDA, HIP, etc. depending on how you compiled it and what is supported on your system.

If you are running on the same CPU as compiling PIConGPU, using `-b omp2b:native` is fine. Then PIConGPU will use OpenMP parallelization on your CPU (regardless if you run 1 or several MPI ranks). Depending on your environment, you may also need to define the standard OpenMP environment variable `$OMP_NUM_THREADS` to expose the desired number of threads to be used. Normally, it should be equal to number of cores you would like to use on each CPU. In case your system has simultaneous multithreading, or Hyperthreading as Intel calls it, the value has to be 2x number of cores to account for that.

**Comment by baobaba13** [comment]
Hi @sbastrakov  Thank you for your reply. I just found that the CPU on my device contains two cores and supports hyper threading. The number of threads is 4. I want to make full use of these four threads and set the mode to `TBG _devices_ {4,4,4},` is this correct? Because the calculation speed is too slow for the example using solid density. Even if I lower the input parameters and calculation range, it still takes several hours to complete the calculation.

**Comment by sbastrakov** [comment]
Sorry if my previous explanation was confusing. No, this would not be a correct way. For a single CPU one should keep `TBG _devices_x = 1`,  `TBG _devices_y = 1`, `TBG _devices_z = 1`. Then with your 2 cores and hyperthreading please set environment variable `OMP_NUM_THREADS` to value 4. Then you will launch PIConGPU as a single MPI process using 4 OpenMP threads, which should generally be best configuration for such a system.

**Comment by baobaba13** [comment]
Hi @sbastrakov 
During the installation of picongpu, do I have any dependencies that include `OpenMP,` or do I need to install `OpenMP` and `MPI` manually? I am No related environment variables are seen in `.bashrc`

**Comment by baobaba13** [comment]
Hi @sbastrakov 
During the installation of picongpu, do I have any dependencies that include `OpenMP,` or do I need to install `OpenMP` and `MPI` manually? I am No related environment variables are seen in `.bashrc`

**Comment by sbastrakov** [comment]
If you mean if PIConGPU somehow automatically installs MPI or OpenMP - no, and it could not do it anyways. Those two are much more basic system things.

MPI is a required dependency of PIConGPU, so you cannot compile it without MPI (and so if you managed to compile PIConGPU your system has some compatible MPI). OpenMP is an optional dependency, but if `-b omp2b:native` (or any `-b omp2b`) works, your system also supports it.

**Comment by baobaba13** [comment]
![5e046ef72b6c7f90a44b90665b61253](https://user-images.githubusercontent.com/39699124/176703679-5cf0fce8-fe0e-4e02-8c07-eba5ececd04a.png)

**Comment by baobaba13** [comment]
Hi @sbastrakov 
According to what you said, I installed MPI when installing picongpu, and my `-b omp2b:native` can also work. I added `export OMP_NUM_THTEADS=4` to `.bashrc` and activated the environment variable. But when running the simulation, the speed is as slow as before. I'm not sure whether the parallel setting is successful. Can you give me some advice? Thank you very much

**Comment by steindev** [comment]
Hey @baobaba13, you can use the commandline tool `htop` while your simulation runs to see how many cores are actually used.

**Comment by baobaba13** [comment]
Hi @steindev 
According to your suggestion, I checked the CPU utilization of my device, as shown in the figure below. It should be that all four threads are in use. I'm not sure whether this speed can be improved, because even so, the simulation speed is still very slow, and I don't know whether this is normal
![NMI@YLVJ_$5SSHE2`WB N6E](https://user-images.githubusercontent.com/39699124/176844962-0dc48e23-b258-49d7-875f-0e25b3d1e24b.png)

**Comment by sbastrakov** [comment]
So if all threads are used it is maybe all one can get for a given machine and simulation setup. Perhaps your initial configuration already did the same and that is why you see no difference from setting `OMP_NUM_THREADS` explicitly.

From your command line I guess you are trying to run a 2d simulation, did you set it up in `dimension.param`? Just setting grid size `z` to 1 does not mean 2d by itself.

**Comment by baobaba13** [comment]
Hi @sbastrakov 
Yes, the example I run is mentioned in the website: [https://picongpu.readthedocs.io/en/latest/examples/Bremsstrahlung/README.html], this is originally a two-dimensional example, The settings in "dimension.param" are shown in the following figure
![TBSOZ$78R9E}RMRK{)9F}CG](https://user-images.githubusercontent.com/39699124/176848383-c8638311-a24d-498e-acd9-07bcd39f01bb.png)

**Comment by steindev** [comment]
@baobaba13 It seems the speed you observe is the best you can obtain for the given setup and machine. I will close this issue. Feel free to open a new issue if you have further questions.
